test/core/xds/xds_client_test.cc

```diff
@@ -3779,6 +3803,367 @@ TEST_F(XdsClientTest, AdsReadWaitsForHandleRelease) {
+  request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"15", /*response_nonce=*/"B",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1", "foo2"});
+  CancelFooWatch(watcher.get(), "foo1", /*delay_unsubscription=*/true);
+  CancelFooWatch(watcher2.get(), "foo2");
+  // Result (remote): The stream to the fallback server has been orphaned.
+  EXPECT_TRUE(stream->Orphaned());
+  EXPECT_THAT(metrics_reporter_->resource_updates_valid(),
+              ::testing::ElementsAre(
+                  ::testing::Pair(
+                      ::testing::Pair(kDefaultXdsServerUrl,
+                                      XdsFooResourceType::Get()->type_url()),
+                      3),
+                  ::testing::Pair(
+                      ::testing::Pair("fallback_xds_server",
+                                      XdsFooResourceType::Get()->type_url()),
+                      2)));
+}
+
+// Test for both servers being unavailable
+TEST_F(XdsClientTest, FallbackReportsError) {
+  FakeXdsBootstrap::FakeXdsServer primary_server(kDefaultXdsServerUrl);
+  FakeXdsBootstrap::FakeXdsServer fallback_server("fallback_xds_server");
+  InitXdsClient(FakeXdsBootstrap::Builder().SetServers(
+      {primary_server, fallback_server}));
+  // Input: Get initial response from primary server.
+  auto watcher = StartFooWatch("foo1");
+  EXPECT_THAT(GetServerConnections(), ::testing::ElementsAre(::testing::Pair(
+                                          kDefaultXdsServerUrl, true)));
+  // Result (local): Resource is delivered to watcher.
+  auto stream = WaitForAdsStream();
+  ASSERT_TRUE(stream != nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  auto request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  // Result (local): Metrics show 1 resource update and 1 cached resource.
+  EXPECT_THAT(GetResourceCounts(),
+              ::testing::ElementsAre(::testing::Pair(
+                  ResourceCountLabelsEq(XdsClient::kOldStyleAuthority,
+                                        XdsFooResourceType::Get()->type_url(),
+                                        "requested"),
+                  1)));
+  TriggerConnectionFailure(primary_server,
+                           absl::UnavailableError("Server down"));
+  stream->MaybeSendStatusToClient(absl::UnavailableError("Server is down"));
+  // Fallback happens now
+  stream = WaitForAdsStream(fallback_server);
+  ASSERT_NE(stream, nullptr);
+  request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  TriggerConnectionFailure(fallback_server,
```

Comment:
After triggering the connection failure, please check that `GetServerConnections()` shows failed connections to both servers.